cin_validator/rules/cin2022_23/rule_4015.py

```diff
@@ -0,0 +1,205 @@
+from typing import Mapping
+
+import pandas as pd
+
+from cin_validator.rule_engine import CINTable, RuleContext, rule_definition
+from cin_validator.test_engine import run_rule
+from cin_validator.utils import make_census_period
+
+# Get tables and columns of interest from the CINTable object defined in rule_engine/__api.py
+
+CINplanDates = CINTable.CINplanDates
+CINPlanStartDate = CINplanDates.CINPlanStartDate
+LAchildID = CINplanDates.LAchildID
+
+CINdetails = CINTable.CINdetails
+CINreferralDate = CINdetails.CINreferralDate
+
+# define characteristics of rule
+@rule_definition(
+    code=4015,
+    # Note that even if multiple tables are involved, one table will be named in the module column.
+    module=CINTable.CINplanDates,
+    message="The CIN Plan start date cannot be before the referral date",
+    affected_fields=[
+        CINPlanStartDate,
+        CINreferralDate,
+    ],
+)
+def validate(
+    data_container: Mapping[CINTable, pd.DataFrame], rule_context: RuleContext
+):
+    """
+    Where present, the <CINPlanStartDate> (N00689) must be on or after the <CINReferralDate> (N00100)
+    """
+    df_cindetail = data_container[CINdetails].copy()
+    df_cinplan = data_container[CINplanDates].copy()
+
+    print(df_cindetail)
+    print(df_cinplan)
+
+    df_cindetail.index.name = "ROW_ID"
+    df_cinplan.index.name = "ROW_ID"
+
+    df_cindetail.reset_index(inplace=True)
+    df_cinplan.reset_index(inplace=True)
+
+    df_cindetail = df_cindetail[
+        ["ROW_ID", "LAchildID", "CINdetailsID", "CINreferralDate"]
+    ]
+    df_cinplan = df_cinplan[["ROW_ID", "LAchildID", "CINdetailsID", "CINPlanStartDate"]]
+
+    df_cinplan = df_cinplan[df_cinplan[CINPlanStartDate].notna()]
+
+    df_merged = df_cindetail.merge(
+        df_cinplan,
+        how="inner",
+        on=["LAchildID", "CINdetailsID"],
+        suffixes=["_det", "_plan"],
+    )
+
+    df_merged.query(r"CINPlanStartDate < CINreferralDate", inplace=True)
+
+    df_merged["ERROR_ID"] = tuple(
+        zip(
+            df_merged[LAchildID],
+            df_merged[CINPlanStartDate],
+            df_merged[CINreferralDate],
+        )
+    )
+
+    df_cindet_issues = (
+        df_cinplan.merge(
+            df_merged, how="inner", left_on="ROW_ID", right_on="ROW_ID_plan"
+        )
+        .groupby("ERROR_ID")["ROW_ID"]
+        .apply(list)
+        .reset_index()
+    )
+
+    df_cindet_issues.sort_values(["ROW_ID"], ignore_index=True, inplace=True)
```

Comment:
This line matters... the " assert issue_rows.equals(expected_df) " near the bottom can fail otherwise.